capstan_test.go

```diff
@@ -22,31 +22,38 @@ func runCapstan(command []string, root string) *exec.Cmd {
 	return c
 }
 
-func TestCommandErrors(t *testing.T) {
+func TestCommandErrorCodes(t *testing.T) {
 	root, err := ioutil.TempDir("", "capstan-root")
 	if err != nil {
 		t.Errorf("capstan: %v", err)
 	}
 	defer os.RemoveAll(root)
 
-	m := map[string]string{
-		"build foo":  "open Capstanfile: no such file or directory\n",
-		"build":      "usage: capstan build [image-name]\n",
-		"pull":       "usage: capstan pull [image-name]\n",
-		"import":     "usage: capstan import [image-name] [image-file]\n",
-		"import foo": "usage: capstan import [image-name] [image-file]\n",
-		"rmi":        "usage: capstan rmi [image-name]\n",
-		"run foo":    "foo: no such image\n",
-		"run":        "No Capstanfile found, unable to run.\n",
-	}
-	for key, value := range m {
-		cmd := runCapstan(strings.Fields(key), root)
-		out, err := cmd.Output()
-		if err != nil {
-			t.Errorf("capstan: %v", err)
+	m := []struct {
+		cmd     string
+		errCode int
+		errMsg  string
+	}{
+		{"build foo", 65, "open Capstanfile: no such file or directory\n"},
+		{"build", 64, "usage: capstan build [image-name]\n"},
+		{"pull", 64, "usage: capstan pull [image-name]\n"},
+		{"import", 64, "usage: capstan import [image-name] [image-file]\n"},
+		{"import foo", 64, "usage: capstan import [image-name] [image-file]\n"},
+		{"rmi", 64, "usage: capstan rmi [image-name]\n"},
+		{"run foo", 65, "foo: no such image\n"},
+		{"run", 65, "Missing Capstanfile or package metadata\n"},
+		{"package help compose", 0, "capstan package compose - composes the package and all its dependencies into OSv imag"},
+	}
+	for _, args := range m {
+		cmd := runCapstan(strings.Fields(args.cmd), root)
+		out, err := cmd.CombinedOutput()
+		if (args.errCode == 0) && (err != nil) ||
+			(args.errCode != 0) && (err == nil) ||
+			(err != nil) && !strings.Contains(err.Error(), fmt.Sprintf("%d", args.errCode)) {
+			t.Errorf("capstan %s: %v", args.cmd, err)
 		}
-		if g, e := string(out), value; g != e {
-			t.Errorf("capstan: want %q, got %q", e, g)
+		if g, e := string(out), args.errMsg; !strings.Contains(g, e) {
+			t.Errorf("capstan %s: want %q, got %q", args.cmd, e, g)
```

Comment:
I haven't actually checked that's true in our case, but check this example: https://play.golang.org/p/IQYPp3tguP. As you can see, `%q` will wrap the text in quotes.

In this case I think this is ok. First `%s` will result in `capstan package compose` (for example), but the other two will produce `want: "what I want", got "are you kidding me"`.